reactive/openstack.py

```diff
@@ -131,7 +131,10 @@ def create_or_update_loadbalancers():
         layer.status.blocked(str(e))
 
 
-@hook('stop')
+@hook("stop")
 def cleanup():
-    # TODO: Also clean up removed LBs as they go away
-    layer.openstack.cleanup()
+    layer.status.maintenance("Cleaning load balancers")
+    for key in layer.openstack.LoadBalancer.get_cached_lbs():
+        lb = layer.openstack.LoadBalancer.load_from_cached(key)
```

Comment:
* `.get_cached_lbs` calls `unitdata.kv.getrange(key_prefix)`, and returns `{key: info}` where `key` does not have the `key_prefix` (i.e. from `my.key.prefix.KEY1`, only `KEY1` is returned in the generated dict.
*  `.load_from_cached` calls `unitdata.kv.get(key)`, but since only `KEY1` is returned, and not the whole key name, I believe it fails, raising an `OpenStackLBError(action="load")`

I think, the `.get_cached_lbs` could return the full key-value information, instead of the list of keys, so the for loop would collect `for key, value in ...` and `.load_from_cached` would return a `LoadBalancer` instance with the specific information (app_name, port, subnet, algorithm, fip_net, manage_secgrps).

Reference
1. https://github.com/juju/charm-helpers/blob/6dc74c68cfef82e85e55eb1f9ab463d503143489/charmhelpers/core/unitdata.py#L199,L227

Reply:
I think you are wrong, because the function `unitdata.kv.getrange` had the parameter `strip=False` and at the [line224](https://github.com/juju/charm-helpers/blob/6dc74c68cfef82e85e55eb1f9ab463d503143489/charmhelpers/core/unitdata.py#L224) you can see the condition, which will change `prefix=''`.

But I like the idea that the `get_cached_lbs` function returns full key-value information that I will use to load `LoadBalancer` as  `LoadBalancer(value['app_name'], ...)`.

I did not notice the [error](https://pastebin.ubuntu.com/p/r9FFH6HCS6/) you mentioned above.